action.yml

```
@@ -62,18 +70,21 @@ runs:
         echo "logs-path=$L_PATH" >> $GITHUB_OUTPUT
       shell: bash
 
-      # MATRIX_STRING is a join of all matrix variables that helps concurrent runs have a unique cache key.
-      # The underscore at the end of the restore key demarks the end of the restore section. Without this
-      # a runner without a matrix has a restore key that will cause impropper clearing of caches from those
-      # with a matrix.
-    - id: keys
+    - name: Generate Keys
+      id: keys
       run: |
-        [ "${{ inputs.include-matrix }}" == "true" ] && MATRIX_STRING="${{ join(matrix.*, '-') }}"
-        [ -n "$MATRIX_STRING" ] && MATRIX_STRING="-${MATRIX_STRING}"
-        RESTORE_KEY="${{ inputs.cache-name }}-${{ runner.os }}${MATRIX_STRING}_"
-        echo "restore-key=${RESTORE_KEY}" >> $GITHUB_OUTPUT
-        echo "key=${RESTORE_KEY}${{ github.run_id }}-${{ github.run_attempt }}" >> $GITHUB_OUTPUT
+        # `matrix_key` joins all of matrix keys/values (including nested objects) to ensure that concurrent runs each use a unique cache key.
+        # When `matrix` isn't set for the job then `MATRIX_JSON=null`.
+        if [ "${{ inputs.include-matrix }}" == "true" ] && [ "$MATRIX_JSON" != "null" ]; then
+          matrix_key=$(echo "$MATRIX_JSON" | jq 'paths(type != "object") as $p | ($p | join("-")) + "=" + (getpath($p) | tostring)' | jq -rs 'join(";") | . + ";"')
+        fi
+        restore_key="${{ inputs.cache-name }};os=${{ runner.os }};${matrix_key}"
```

I think we can drop `runner.os` if a user wants multiple OS support they'd either add the OS into the `matrix` or they would need to use a different workflow name or job name:
```suggestion
        restore_key="${{ inputs.cache-name }};${matrix_key}"
```

---

Force including `runner.os` is mostly a failsafe. I generally agree but I worried about some edge case configuration

---

I'm not sure I follow this point
> workflow name and job name to no longer be specified in the cache-name

WIthout it as a default, if two workflows/jobs have the same matrix then they'd share the same cache key

---

I'm advocating for the workflow/job to not be a part of the `cache-name` default but rather be a mandatory part of this step. I originally did that in this PR but discovered that the tests for this action require require these values to be overridden so that we can have separate save/restore jobs.

---

Ok, but forcing them into the cache key means users can't use caches across different workflow/jobs if they wanted to (without having to manually override the full cache-key as you propose in https://github.com/julia-actions/cache/issues/89)

---

Very true. I think having the option of supplying a full `cache-key` and also returning the generated cache key as an output may allow the use of caches across different jobs and still making these portions be mandatory. I appreciate you bringing up these alternate use cases so I have a clearer understanding on some implicit design of this action.

---

Yeah, though I didn't consolidate it down to a clear list, I did try to think through a user specification here, the key points that come to mind being:
- not breaking existing usages
- being sensibly overridable for different approaches
- working for standard julia CI approaches
- being fail-safe for `/compiled` cache issues wrt. platform

Also bear in mind that the cache-key has the workflow/job in it, but the lookup key doesn't. That means it's possible to restore without paying attention to workflow/job, but will always be able to save. (it'd be nice if `actions/cache` can force save, but it can't)